java.lang.IllegalArgumentException: End gap 8 is too big: capacity is 7

## Actual Behavior

```
Exception in thread "DefaultDispatcher-worker-3" java.lang.IllegalArgumentException: End gap 8 is too big: capacity is 7
	at io.ktor.utils.io.core.BufferKt.endGapReservationFailedDueToCapacity(Buffer.kt:463)
	at io.ktor.utils.io.core.Buffer.reserveEndGap(Buffer.kt:220)
	at io.ktor.utils.io.core.AbstractOutput.appendNewChunk(AbstractOutput.kt:195)
	at io.ktor.utils.io.core.AbstractOutput.prepareWriteHead(AbstractOutput.kt:497)
	at io.ktor.utils.io.core.OutputPrimitivesKt.writeIntFallback(OutputPrimitives.kt:133)
	at io.ktor.utils.io.core.OutputPrimitivesKt.writeInt(OutputPrimitives.kt:22)
	at io.rsocket.kotlin.frame.Frame.toPacket(Frame.kt:41)
	at io.rsocket.kotlin.ConnectionKt.sendFrame(Connection.kt:42)
	at io.rsocket.kotlin.internal.RSocketState$start$3.invokeSuspend(RSocketState.kt:197)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.selects.SelectBuilderImpl.resumeWith(Select.kt:302)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
java.util.concurrent.CancellationException: Connection closed
	at kotlinx.coroutines.ExceptionsKt.CancellationException(Exceptions.kt:22)
	at io.rsocket.kotlin.internal.RSocketState$start$2.invoke(RSocketState.kt:183)
	at io.rsocket.kotlin.internal.RSocketState$start$2.invoke(RSocketState.kt:33)
	at kotlinx.coroutines.InvokeOnCompletion.invoke(JobSupport.kt:1412)
	at kotlinx.coroutines.JobSupport.notifyCompletion(JobSupport.kt:1544)
	at kotlinx.coroutines.JobSupport.completeStateFinalization(JobSupport.kt:326)
	at kotlinx.coroutines.JobSupport.finalizeFinishingState(JobSupport.kt:241)
	at kotlinx.coroutines.JobSupport.continueCompleting(JobSupport.kt:951)
	at kotlinx.coroutines.JobSupport.access$continueCompleting(JobSupport.kt:29)
	at kotlinx.coroutines.JobSupport$ChildCompletion.invoke(JobSupport.kt:1175)
	at kotlinx.coroutines.JobSupport.notifyCompletion(JobSupport.kt:1544)
	at kotlinx.coroutines.JobSupport.completeStateFinalization(JobSupport.kt:326)
	at kotlinx.coroutines.JobSupport.finalizeFinishingState(JobSupport.kt:241)
	at kotlinx.coroutines.JobSupport.tryMakeCompletingSlowPath(JobSupport.kt:922)
	at kotlinx.coroutines.JobSupport.tryMakeCompleting(JobSupport.kt:875)
	at kotlinx.coroutines.JobSupport.makeCompletingOnce$kotlinx_coroutines_core(JobSupport.kt:840)
	at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:111)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at kotlinx.coroutines.selects.SelectBuilderImpl.resumeWith(Select.kt:302)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
Caused by: java.lang.IllegalArgumentException: End gap 8 is too big: capacity is 7
	at io.ktor.utils.io.core.BufferKt.endGapReservationFailedDueToCapacity(Buffer.kt:463)
	at io.ktor.utils.io.core.Buffer.reserveEndGap(Buffer.kt:220)
	at io.ktor.utils.io.core.AbstractOutput.appendNewChunk(AbstractOutput.kt:195)
	at io.ktor.utils.io.core.AbstractOutput.prepareWriteHead(AbstractOutput.kt:497)
	at io.ktor.utils.io.core.OutputPrimitivesKt.writeIntFallback(OutputPrimitives.kt:133)
	at io.ktor.utils.io.core.OutputPrimitivesKt.writeInt(OutputPrimitives.kt:22)
	at io.rsocket.kotlin.frame.Frame.toPacket(Frame.kt:41)
	at io.rsocket.kotlin.ConnectionKt.sendFrame(Connection.kt:42)
	at io.rsocket.kotlin.internal.RSocketState$start$3.invokeSuspend(RSocketState.kt:197)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	... 7 more
```

## Steps to Reproduce

With 

./rsocket-cli ws://rsocket-demo-1.ew.r.appspot.com/rsocket
./rsocket-cli wss://rsocket-demo-1.ew.r.appspot.com/rsocket

## Your Environment

GraalVM (11) or JDK11

--- a/rsocket-transport-ktor/src/commonMain/kotlin/io/rsocket/kotlin/transport/ktor/WebSocketConnection.kt
+++ b/rsocket-transport-ktor/src/commonMain/kotlin/io/rsocket/kotlin/transport/ktor/WebSocketConnection.kt
@@ -27,11 +27,16 @@ public class WebSocketConnection(private val session: WebSocketSession) : Connec
     override val job: CompletableJob = Job(session.coroutineContext[Job])
 
     override suspend fun send(packet: ByteReadPacket) {
-        session.send(Frame.Binary(true, packet))
+        session.send(packet.readBytes())
     }
 
     override suspend fun receive(): ByteReadPacket {
-        return ByteReadPacket(session.incoming.receive().data)
+        val frame = session.incoming.receive()
+        // TODO replace with ByteReadPacket(frame.data) after fix for KTOR-960
+        //  now it's not zero copy
+        return buildPacket {
+            writeFully(frame.data)
+        }
     }
 
 }

--- a/rsocket-transport-ktor/src/jvmTest/kotlin/io/rsocket/kotlin/WebSocketConnectionTest.kt
+++ b/rsocket-transport-ktor/src/jvmTest/kotlin/io/rsocket/kotlin/WebSocketConnectionTest.kt
@@ -1,0 +1,100 @@
+/*
+ * Copyright 2015-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.rsocket.kotlin
+
+import io.ktor.application.*
+import io.ktor.client.*
+import io.ktor.routing.*
+import io.ktor.server.engine.*
+import io.rsocket.kotlin.core.*
+import io.rsocket.kotlin.keepalive.*
+import io.rsocket.kotlin.payload.*
+import io.rsocket.kotlin.test.*
+import io.rsocket.kotlin.transport.ktor.client.*
+import io.rsocket.kotlin.transport.ktor.server.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.*
+import kotlin.random.*
+import kotlin.test.*
+import kotlin.time.*
+import io.ktor.client.engine.cio.CIO as ClientCIO
+import io.ktor.client.features.websocket.WebSockets as ClientWebSockets
+import io.ktor.server.cio.CIO as ServerCIO
+import io.ktor.websocket.WebSockets as ServerWebSockets
+import io.rsocket.kotlin.transport.ktor.client.RSocketSupport as ClientRSocketSupport
+import io.rsocket.kotlin.transport.ktor.server.RSocketSupport as ServerRSocketSupport
+
+class WebSocketConnectionTest : SuspendTest {
+    private val port = Random.nextInt(20, 90) * 100
+    private val client = HttpClient(ClientCIO) {
+        install(ClientWebSockets)
+        install(ClientRSocketSupport) {
+            connector = RSocketConnector {
+                connectionConfig {
+                    keepAlive = KeepAlive(500.milliseconds)
+                }
+            }
+        }
+    }
+
+    private var responderJob: Job? = null
+
+    private val server = embeddedServer(ServerCIO, port) {
+        install(ServerWebSockets)
+        install(ServerRSocketSupport)
+        install(Routing) {
+            rSocket {
+                RSocketRequestHandler {
+                    requestStream {
+                        it.release()
+                        flow {
+                            var i = 0
+                            while (true) {
+                                emit(buildPayload { data((++i).toString()) })
+                                delay(1000)
+                            }
+                        }
+                    }
+                }.also { responderJob = it.job }
+            }
+        }
+    }
+
+    override suspend fun before() {
+        server.start()
+        delay(1000)
+    }
+
+    override suspend fun after() {
+        server.stop(0, 0)
+    }
+
+    @Test
+    fun testWorks() = test {
+        val rSocket = client.rSocket(port = port)
+        val requesterJob = rSocket.job
+
+        rSocket
+            .requestStream(Payload.Empty)
+            .take(2)
+            .onEach { delay(100); it.release() }
+            .collect()
+
+        assertTrue(requesterJob.isActive)
+        assertTrue(responderJob?.isActive!!)
+    }
+}

